tests/test_extension.py

```diff
@@ -0,0 +1,683 @@
+        mock_logger.info.assert_called_with('ok_pre_invocation')
+
+    def test_two_extensions_on_same_trigger(self):
+        """Test if two extensions can be registered on the same trigger
+        """
+        class NewExtension1(FuncExtensionBase):
+            def __init__(self, trigger_name: str):
+                super().__init__(trigger_name)
+
+            def pre_invocation(self, logger: Logger, context: Context,
+                               *args, **kwargs) -> None:
+                logger.info('ok_before_1')
+
+        class NewExtension2(FuncExtensionBase):
+            def __init__(self, trigger_name: str):
+                super().__init__(trigger_name)
+
+            def pre_invocation(self, logger: Logger, context: Context,
+                               *args, **kwargs) -> None:
+                logger.info('ok_before_2')
+
+        # Check if both extensions are registered under the same hook
+        NewExtension1(self.mock_file_path)
+        NewExtension2(self.mock_file_path)
+        hooks = ExtensionMeta.get_function_hooks('HttpTrigger')
+
+        # Check if the before invocation hook matches metadata
+        extension_names = map(
+            lambda x: getattr(x, 'ext_name'),
+            hooks.pre_invocation
+        )
+        self.assertIn('NewExtension1', extension_names)
+        self.assertIn('NewExtension2', extension_names)
+
+    def test_backward_compatilbility_less_arguments(self):
+        """Test if the existing extension implemented the interface with
+        less arguments
+        """
+        class ExtensionWithLessArgument(FuncExtensionBase):
+            def __init__(self, file_path: str):
+                super().__init__(file_path)
+                self.executed = False
+
+            # Drop arguments
+            def pre_invocation(self):
+                self.executed = True
+
+        # Check if the before invocation hook matches metadata
+        ext_instance = ExtensionWithLessArgument(self.mock_file_path)
+        hooks = ExtensionMeta.get_function_hooks('HttpTrigger')
+
+        # Check if implementation works
+        hook_meta = hooks.pre_invocation[0]
+        self.assertEqual(hook_meta.ext_name, 'ExtensionWithLessArgument')
+
+        # Check if the hook implementation executes
+        hook_meta.ext_impl()
+        self.assertTrue(ext_instance.executed)
+
+
+class TestAppExtensionBase(unittest.TestCase):
+
+    def setUp(self):
+        self.patch_os_environ = patch.dict('os.environ', os.environ.copy())
+        self.patch_os_environ.start()
+
+    def tearDown(self) -> None:
+        self.patch_os_environ.stop()
+        ExtensionMeta._info.clear()
+        ExtensionMeta._func_exts.clear()
+        ExtensionMeta._app_exts = None
+
+    def test_empty_extension_should_pass(self):
+        """An application extension can be registered directly since it never
+        gets instantiate
+        """
+        class NewAppExtension(AppExtensionBase):
+            pass
```

Comment:
So - defining `NewAppExtension` should not raise an exception. Is this what you are testing?

Reply:
Yes, amended the pydoc to demonstrate this test.